Fix: if field has custom decoder, schema takes it into account

deansg opened the pull request:
If a field has a custom decoder, replace its _deserialize logic with a dummy method, as it will be taken care of by the decoder.

Attempts to solve #460

deansg commented:
This is just an initial version. I'd like to know if I'm in the right direction or if this isn't it.

deansg commented:
Regarding the code quality check, that's a known mypy issue as seen here:
https://github.com/python/mypy/issues/2427

We can just add `--disable-error-code=method-assign` to the check if we want

george-zubrienko commented on dataclasses_json/mm.py:
```diff
@@ -162,15 +163,16 @@ def __init__(self, *args, **kwargs):
 
         @typing.overload
-        def dump(self, obj: typing.List[A], many: typing.Optional[bool] = None) -> typing.List[TEncoded]:  # type: ignore
+        def dump(self, obj: typing.List[A],  # type: ignore
```
Please revert code formatting changes, so the PR contains functional changes only :)

george-zubrienko commented on dataclasses_json/mm.py:
```diff
@@ -318,13 +320,26 @@ def schema(cls, mixin, infer_missing):
                 options['data_key'] = metadata.letter_case(field.name)
 
             t = build_type(type_, options, mixin, field, cls)
+            _account_for_decoder_if_necessary(field, t)
```
```suggestion
            if field.metadata.get('dataclasses_json', {}).get('decoder'):
                t._deserialize = lambda v, *_a, **_kw: v
```

Also maybe add comment why a private field is being mutated here?

george-zubrienko commented on dataclasses_json/mm.py:
```diff
@@ -318,13 +320,26 @@ def schema(cls, mixin, infer_missing):
     return schema
 
 
+def _account_for_decoder_if_necessary(field: dataclasses.Field, t: fields.Field):
+    decoder = field.metadata.get('dataclasses_json', {}).get('decoder')
+    if decoder:
+        # Used in order to avoid mypy error. See https://github.com/python/mypy/issues/2427 for more details
+        t._deserialize = _dummy_deserialize  # type: ignore
+
+
+def _dummy_deserialize(value: typing.Any, *_args, **_kwargs):
+    return value
+
+
```
```suggestion
```

george-zubrienko reviewed (commented):
Thank you for contribution!

Minor comments, I don't think the approach is wrong, but I'm also not very string on marshmellow side, so a bit of explanation of this solution would be nice :)

deansg commented:
@george-zubrienko I reverted all formatting changes except for a single line which only contained spaces which IntelliJ just won't let me change at all, I hope it's not a deal breaker.
I also added an extended comment explaining the change.
Let me know if you have more comments.

george-zubrienko reviewed (approved):
Awesome, nice and clear! Thanks a lot for fixing this bug!
